packages/@react-spectrum/textfield/src/TextArea.tsx

```diff
@@ -26,31 +27,94 @@ function TextArea(props: SpectrumTextFieldProps, ref: RefObject<TextFieldRef>) {
   } = props;
 
+  let fallbackRef = useRef<TextFieldRef>(null);
+  if (!ref) {
+    ref = fallbackRef;
+  }
+
   // not in stately because this is so we know when to re-measure, which is a spectrum design
   let [inputValue, setInputValue] = useControlledState(props.value, props.defaultValue, () => {});
 
   let inputRef = useRef<HTMLTextAreaElement>();
 
+  // Store styleProps to apply styles appropriately on height change.
+  let {styleProps} = useStyleProps(props);
+  let initialHeightsRef = useRef(null);
+
   let onHeightChange = useCallback(() => {
+    let input = inputRef.current;
+    let inputContainer = input.parentElement;
+    let field = ref.current.UNSAFE_getDOMNode();
+
+    // current offsetHeight of the textarea
+    let inputOffsetHeight = input.offsetHeight;
+
+    // Label or helpText height might change, so we have to recalculate
+    let labelHeight = labelPosition !== 'side' && !field.firstElementChild.contains(input) ? (field.firstElementChild as HTMLElement).offsetHeight : 0;
+    
+    // With labelPosition === 'side', the TextArea and HelpText will be contained within a wrapper.
+    let fieldWrapper = labelPosition === 'side' ? field.lastElementChild : field;
+    let helpTextHeight = !fieldWrapper.lastElementChild.contains(input) ? (fieldWrapper.lastElementChild as HTMLElement).offsetHeight : 0;
+
+    // store a ref containing the initial size values for comparison
+    if (initialHeightsRef.current === null) {
+      initialHeightsRef.current = {
+        inputHeight: inputOffsetHeight,
+        fieldHeight: field.offsetHeight
+      };
+    }
+    let {inputHeight, fieldHeight} = initialHeightsRef.current;
+    let calculatedFieldHeight = Math.max(inputHeight, inputOffsetHeight) + labelHeight + helpTextHeight;
     if (isQuiet) {
-      let input = inputRef.current;
       let prevAlignment = input.style.alignSelf;
       input.style.alignSelf = 'start';
       input.style.height = 'auto';
-      input.style.height = `${input.scrollHeight}px`;
+      input.style.height = `${Math.max(input.scrollHeight, inputHeight)}px`;
       input.style.alignSelf = prevAlignment;
+      inputContainer.style.flexGrow = '0';
+      for (const [key, value] of Object.entries(styleProps.style)) {
+        switch (key) {
+          case 'height':
+          case 'maxHeight':
+            field.style[key] = calculatedFieldHeight <= fieldHeight ? value : '';
+            break;
+        }
+      }
+    } else {  
+      for (const [key, value] of Object.entries(styleProps.style)) {
+        switch (key) {
+          case 'height':
+            field.style[key] = calculatedFieldHeight <= fieldHeight ? value : `${calculatedFieldHeight}px`;
+            
+            // When the textarea resizes smaller than the initial height, 
+            // its container should shrink, so that any helpText will align 
+            // to the bottom of the textarea.
+            inputContainer.style.flexGrow = inputOffsetHeight < inputHeight ? '0' : '';
```

Comment:
NOTE: Just some musings, I'm perfectly fine with having this PR go in as is since we'll eventually revisit the layout for TextArea  but figured I'd bring it up if ppl had opinions.

Noticed that the field height doesn't shrink past its user provided height:
![image](https://user-images.githubusercontent.com/9661127/141019150-0ab6e0a7-0acb-4066-93ff-a43cb5cfc239.png)

I'm a bit torn on what the behavior should be here. The current behavior matches what we had in the past (aka the field height stays the same if the user sets an explicit height) but since we are now changing it so the field height grows when the textarea is enlarged via user drag event, perhaps we should also have the field height shrink when the user makes the textarea smaller than the applied `height`? Could be done via something like this perhaps:
```
            if (inputOffsetHeight < inputHeight) {
              field.style[key] = `${helpTextHeight + inputOffsetHeight + labelHeight}px`
            }
```
Then again, perhaps we don't want that behavior, might be weird. To be honest, perhaps `height` should be `initialHeight` if anything

Reply:
I definitely wrestled with this problem, my question being if we always resize to the current height, will the user specified height ever be meaningful?

Reply:
if it has a user provided height, maybe we should just disable resizing?